source/D2CommonDefinitions/include/D2PacketDef.h

```diff
@@ -296,14 +296,19 @@ struct D2GSPacketClt31		//size of 0x09
 	int32_t unk0x01[2];			//0x01
 };
 
+enum D2TransactionMode : uint32_t {
+    TRANSACTIONMODE_BUY = 0,
+    TRANSACTIONMODE_GAMBLE = 2,
+    TRANSACTIONMODE_FILL = 0x80000000
+};
+
 struct D2GSPacketClt32		//size of 0x11
 {
-	uint8_t nHeader;			//0x00
-	int32_t unk0x01;			//0x01
-	int32_t unk0x05;			//0x05
-	uint16_t unk0x09;			//0x09
-	uint16_t unk0x0B;			//0x0B
-	int32_t unk0x0D;			//0x0D
+	uint8_t nHeader;						//0x00
+	int32_t dwUnitGUID;						//0x01
+	int32_t dwItemGUID;						//0x05
+	D2TransactionMode dwTransactionMode;	//0x09
```

Comment:
Given:
```
return D2GAME_NPC_BuyItemHandler_6FCC92A0(pGame, pUnit, *(int32_t*)((char*)pPacket + 1), *(int32_t*)((char*)pPacket + 5), (*(int32_t*)((char*)pPacket + 9) & INT_MAX) >> 16, *(int32_t*)((char*)pPacket + 9), *(int32_t*)((char*)pPacket + 13), *(int32_t*)((char*)pPacket + 9) & 0x80000000);
```
This field is a bit more interesting, as sometimes it reads just the top bit, sometimes, the 15 upper bits minus the top one, sometimes the whole thing, however I think making struct/union/bitstruct for this might be annoying.

Reply:
I think we may want to have a union or at least helper functions in this case. After a quick look I think those 4 bytes pack the following information:

- Transaction type bits 0 - 15 (`enum` should be in D2Common, as it must be used in `ITEMS_CalculateTransactionCost`)
- Vendor price reduction bits 16-30
- Fill item (book, pots, scrolls...) bit 31

Reply:
Transaction types are:
- 0 buy
- 1 sell
- 2 gamble
- 3 repair

Reply:
Not sure that is correct, as its a buy packet, so sell doesn't make sense. I also assumed there is a separate repair packet.

Reply:
Well, to be fair, values 1 and 3 are unused, but the field is converted and used as transaction type, which may hold any of the 4 values mentioned above.
I suppose we could have a specific enum for the packet, then cast to the real transaction type enum.

Reply:
As for the issue of the packing, I just checked and it's packed because they use the same function for all packets of size 0x11 (`D2Client.0x6FAADAA0`) which takes 4 DWORD params.

For reference here is how it is computed:

```c++
        nPackedTransationInfo = (unsigned __int16)nItemAnimMode << 16;
        if ( gbVendorGamble_6FBB5D7C )
            nPackedTransationInfo |= 2u;
        if ( (nKeys & 4) != 0 )                 // is Shift held
        {
            nClassId = pItem ? pItem->dwClassId : -1;
            pItemRecord = D2COMMON_10600_GetItemTxtRecord(nClassId);
            if ( pItemRecord )
            {
                if ( pItemRecord->nMultibuy )
                    nPackedTransationInfo |= 0x80000000;
            }
        }
        D2CLIENT_PACKETS_SendPacketSize17_6FAADAA0(
            0x32,
            gnActiveNpcGUID_6FBB5CF5,
            dword_13E25BD5,
            nPackedTransationInfo,
            a5);
```